feat(toggle): fixes disabled issue and adds test

AlexanderMelox opened the pull request:
### Closes #13956 

**Before**
https://github.com/carbon-design-system/carbon/assets/12755042/071a7579-fc21-4b48-b101-2ef6382667a6

**After**


https://github.com/carbon-design-system/carbon/assets/12755042/e4438e1e-5109-413d-bc64-2af52d3a0e21



This PR fixes the issue where you are able to still toggle when the Toggle is disabled.

#### Changelog

**Changed**

- packages/react/src/components/Toggle/Toggle.js
- packages/react/src/components/Toggle/Toggle-test.js

#### Testing / Reviewing

Checked on storybook if you are able to toggle while disabled. Added test to make sure you are not able to toggle when disabled.

AlexanderMelox committed:
feat(toggle): fixes disabled issue and adds test

tw15egan commented on packages/react/src/components/Toggle/Toggle.js:
```diff
@@ -37,7 +38,7 @@ export function Toggle({
 
   function handleClick(e) {
-    if (!readOnly) {
+    if (!readOnly && !disabled) {
```
We probably don't want to fire the `onClick` when disabled, either, since it wouldn't match keyboard parity (can't be focused).

Should we just add `!disabled` to `L87`?

```jsx
if (buttonElement.current && e.target !== buttonElement.current && !disabled) {
```

AlexanderMelox replied:
So I tried adding it there but it's still showing the styles, I wonder if it has to do with CSS :disabled state

AlexanderMelox replied:
Got it, I'm figuring out the selectors now

tw15egan commented:
I'm also seeing some focus styles being applied when clicking on the `disabled` `Toggle`.

I was able to remove those styles locally by changing the selector to this on `L77` in `_toggle.scss`

```scss
  .#{$prefix}--toggle__button:not(:disabled):focus
    + .#{$prefix}--toggle__label
    .#{$prefix}--toggle__switch,
  .#{$prefix}--toggle__button:not(:disabled):active
    + .#{$prefix}--toggle__label
    .#{$prefix}--toggle__switch,
  .#{$prefix}--toggle:not(.#{$prefix}--toggle--disabled):active
    .#{$prefix}--toggle__switch {
```

Seems like the `HCM` fix at the bottom of the file uses the same selector, so should update that block as well.

AlexanderMelox committed:
fix: changes location of disabled check

AlexanderMelox committed:
feat: removes styles that show focus state when disabled

AlexanderMelox commented:
> I'm also seeing some focus styles being applied when clicking on the `disabled` `Toggle`.
> 
> I was able to remove those styles locally by changing the selector to this on `L77` in `_toggle.scss`
> 
> ```css-scss
>   .#{$prefix}--toggle__button:not(:disabled):focus
>     + .#{$prefix}--toggle__label
>     .#{$prefix}--toggle__switch,
>   .#{$prefix}--toggle__button:not(:disabled):active
>     + .#{$prefix}--toggle__label
>     .#{$prefix}--toggle__switch,
>   .#{$prefix}--toggle:not(.#{$prefix}--toggle--disabled):active
>     .#{$prefix}--toggle__switch {
> ```
> 
> Seems like the `HCM` fix at the bottom of the file uses the same selector, so should update that block as well.

I went ahead and added the changes. I removed the CSS that shows focus and active states when the toggle is disabled. In terms of keyboard navigation should the toggle still have focus even thought it's disabled?

AlexanderMelox committed:
Merge branch 'main' into 13956--toggle-disabled-bug

AlexanderMelox committed:
Merge branch 'main' into 13956--toggle-disabled-bug

AlexanderMelox commented:
@tw15egan It's just focusing the selector tab at the bottom! So the disabled Toggle is not in the Tab Order :) 

https://github.com/carbon-design-system/carbon/assets/12755042/1160c4a0-9dd4-4d6f-8de0-403adca9e3d7

francinelucca reviewed (changes requested):
LGTM but seems like 1 tab stop is getting lost, I have to tab twice from the top panel to get to the bottom and it's not clear where the focus is on that "ghost" tab stop 

https://github.com/carbon-design-system/carbon/assets/40550942/51fe0d30-a8d0-4b58-89e2-86b75d936e98

tw15egan commented:
![2023-06-09 10 42 43](https://github.com/carbon-design-system/carbon/assets/11928039/db6319d6-eae4-4060-ac97-76130aeb0919)


Yeah I also tried to capture it, its getting lost somewhere

AlexanderMelox commented:
I'm trying to see where the focus can go, we did disabled this so idk if this is enough. 

<img width="939" alt="Screenshot 2023-06-09 at 10 52 11 AM" src="https://github.com/carbon-design-system/carbon/assets/12755042/007b9c70-8864-43b3-8a68-626357ce5c7c">

AlexanderMelox commented:
<img width="564" alt="Screenshot 2023-06-09 at 10 54 47 AM" src="https://github.com/carbon-design-system/carbon/assets/12755042/30965804-065a-4486-90e8-5d755
<img width="167" alt="Screenshot 2023-06-09 at 10 54 57 AM" src="https://github.com/carbon-design-system/carbon/assets/12755042/01c27734-a548-42c0-9a38-4ce825b5cba9">
daa3536">

So i did some debugging and it looks like focus is still being applied while disabled, we just don't see the styles

tw15egan commented:
@AlexanderMelox weird, even setting the `tabIndex` on the `div` (and/or `button`) it still stays in the tab order 🤔

tw15egan committed:
test(Toggle): fix lint error

AlexanderMelox commented:
The tests are oddly failing

tw15egan committed:
chore(Toggle): add screen to import

kodiakhq[bot] committed:
Merge branch 'main' into 13956--toggle-disabled-bug

tw15egan committed:
Merge branch 'main' into 13956--toggle-disabled-bug
